metricflow/plan_conversion/dataflow_to_sql.py

```diff
@@ -1297,3 +1301,64 @@ def visit_semi_additive_join_node(self, node: SemiAdditiveJoinNode) -> SqlDataSe
                 order_bys=(),
             ),
         )
+
+    def visit_join_to_time_spine_node(self, node: JoinToTimeSpineNode[SourceDataSetT]) -> SqlDataSet:  # noqa: D
+        parent_data_set = node.parent_node.accept(self)
+        parent_alias = self._next_unique_table_alias()
+
+        # Build time spine dataset
+        metric_time_dimension_instance: Optional[TimeDimensionInstance] = None
+        for instance in parent_data_set.metric_time_dimension_instances:
+            if len(instance.spec.identifier_links) == 0:
+                metric_time_dimension_instance = instance
+                break
+        assert (
+            metric_time_dimension_instance
+        ), "Can't query offset metric without a time dimension. Validations should have prevented this."
+        metric_time_dimension_column_name = self.column_association_resolver.resolve_time_dimension_spec(
+            metric_time_dimension_instance.spec
+        ).column_name
+        time_spine_alias = self._next_unique_table_alias()
+        time_spine_dataset = _make_time_spine_data_set(
+            metric_time_dimension_instance=metric_time_dimension_instance,
+            metric_time_dimension_column_name=metric_time_dimension_column_name,
+            time_spine_source=self._time_spine_source,
+            time_spine_table_alias=time_spine_alias,
+            time_range_constraint=node.time_range_constraint,
+        )
+
+        # Build join expression
+        left_expr: SqlExpressionNode = SqlColumnReferenceExpression(
+            col_ref=SqlColumnReference(table_alias=time_spine_alias, column_name=metric_time_dimension_column_name)
+        )
+        if node.offset_window:
+            left_expr = SqlTimeDeltaExpression(
+                arg=left_expr, count=node.offset_window.count, granularity=node.offset_window.granularity
+            )
+        elif node.offset_to_grain_to_date:
+            left_expr = SqlDateTruncExpression(time_granularity=node.offset_to_grain_to_date, arg=left_expr)
+        join_description = SqlJoinDescription(
+            right_source=parent_data_set.sql_select_node,
+            right_source_alias=parent_alias,
+            on_condition=SqlComparisonExpression(
+                left_expr=left_expr,
+                comparison=SqlComparison.EQUALS,
+                right_expr=SqlColumnReferenceExpression(
+                    col_ref=SqlColumnReference(table_alias=parent_alias, column_name=metric_time_dimension_column_name)
+                ),
+            ),
+            join_type=SqlJoinType.LEFT_OUTER,  # TODO: test other join types
+        )
+
+        return SqlDataSet(
+            instance_set=InstanceSet.merge([time_spine_dataset.instance_set, parent_data_set.instance_set]),
+            sql_select_node=SqlSelectStatementNode(
+                description=node.description,
+                select_columns=parent_data_set.sql_select_node.select_columns,
```

Comment:
See commentary on SQL below, but I think we need two changes here:

1. This has to be built out using `create_select_columns_for_instance_sets` like in the `visit_join_to_base_output_node` method. We first need to map the time_spine_alias and parent_alias values to the columns we want to include in the select and then let this update those aliases accordingly.
2. I'm less certain of this, but I think for the offsetting to work we'll want to make sure the output metric_time column is the one from the time_spine, not the metric dataset. You can do that as part of this alias mapping process or in a custom transform on the instance set. I might be wrong, though, it's really hard to be sure without seeing the input/output data.

Reply:
Updated to fix both!